src/go/configgenerator/filter_generator.go

```diff
@@ -0,0 +1,262 @@
+			},
+		})
+	}
+
+	// Add JWT Authn filter if needed.
+	if !serviceInfo.Options.SkipJwtAuthnFilter {
+		// TODO(b/176432170): Handle errors here, prevent startup.
+		filterGenerators = append(filterGenerators, &FilterGenerator{
+			FilterName:            util.JwtAuthn,
+			FilterGenFunc:         jaFilterGenFunc,
+			PerRouteConfigGenFunc: jaPerRouteFilterConfigGen,
+		})
+	}
+
+	// Add Service Control filter if needed.
+	if !serviceInfo.Options.SkipServiceControlFilter {
+		filterGenerators = append(filterGenerators, &FilterGenerator{
+			FilterName:            util.ServiceControl,
+			FilterGenFunc:         scFilterGenFunc,
+			PerRouteConfigGenFunc: scPerRouteFilterConfigGen,
+		})
+	}
+
+	// Add gRPC Transcoder filter and gRPCWeb filter configs for gRPC backend.
+	if serviceInfo.GrpcSupportRequired {
+		// grpc-web filter should be before grpc transcoder filter.
+		// It converts content-type application/grpc-web to application/grpc and
+		// grpc transcoder will bypass requests with application/grpc content type.
+		// Otherwise grpc transcoder will try to transcode a grpc-web request which
+		// will fail.
+		filterGenerators = append(filterGenerators, &FilterGenerator{
+			FilterName: util.GRPCWeb,
+			FilterGenFunc: func(sc *ci.ServiceInfo) (*hcmpb.HttpFilter, []*ci.MethodInfo, error) {
+				return &hcmpb.HttpFilter{
+					Name: util.GRPCWeb,
+				}, nil, nil
+			},
+		})
+
+		filterGenerators = append(filterGenerators, &FilterGenerator{
+			FilterName: util.GRPCJSONTranscoder,
+			FilterGenFunc: func(sc *ci.ServiceInfo) (*hcmpb.HttpFilter, []*ci.MethodInfo, error) {
+				return makeTranscoderFilter(serviceInfo), nil, nil
+			},
+		})
+	}
+
+	filterGenerators = append(filterGenerators, &FilterGenerator{
+		FilterName:            util.BackendAuth,
+		FilterGenFunc:         baFilterGenFunc,
+		PerRouteConfigGenFunc: baPerRouteFilterConfigGen,
+	})
+
+	filterGenerators = append(filterGenerators, &FilterGenerator{
+		FilterName:            util.PathRewrite,
+		FilterGenFunc:         prFilterGenFunc,
+		PerRouteConfigGenFunc: prPerRouteFilterConfigGen,
+	})
+
+	if serviceInfo.Options.EnableGrpcForHttp1 {
+		// Add GrpcMetadataScrubber filter to retain gRPC trailers
+
+		filterGenerators = append(filterGenerators, &FilterGenerator{
+			FilterName: util.GrpcMetadataScrubber,
+			FilterGenFunc: func(sc *ci.ServiceInfo) (*hcmpb.HttpFilter, []*ci.MethodInfo, error) {
+				return &hcmpb.HttpFilter{
+					Name: util.GrpcMetadataScrubber,
+				}, nil, nil
+			},
+		})
+	}
+
+	// Add Envoy Router filter so requests are routed upstream.
+	// Router filter should be the last.
+	filterGenerators = append(filterGenerators, &FilterGenerator{
+		FilterName: util.Router,
+		FilterGenFunc: func(sc *ci.ServiceInfo) (*hcmpb.HttpFilter, []*ci.MethodInfo, error) {
+			return makeRouterFilter(serviceInfo.Options), nil, nil
+		},
+	})
+	return filterGenerators, nil
+}
+
+func makeTranscoderFilter(serviceInfo *ci.ServiceInfo) *hcmpb.HttpFilter {
```

Comment:
can we move these into different files as well, one per filter?

My idea is similar to Wayne's. In g3, we will have an override of `MakeFilterGenerators`. So the `filter_generator.go` file will have different contents in g3 vs OSS. All the other per-filter files can be synced as needed into g3.

OSS:
```
filter_generator.go
sc_v1.go
backend_auth.go
transcoding.go
...
```

g3
```
filter_generator.go (modified contents for g3 filters)
sc_v2.go
transcoding.go
...
```

So we do not want to tie any filter-specific functionality into filter_generator.go.

Reply:
Let me know if you have a different idea for how filters will be generated in google3

Reply:
Yes, I like the idea of each filter has its own .go file,  or even its own go package.   There are different between Teju idea and mime.

Mine:  we don't have separate MakeFilterGenerator,  one for OSS and one for g3.  There is only one, a very simple, it just read the registered filter_list.  Similar to Envoy extensions,  there is not MakeExtersions function,   each extension just register itself.  By removing, adding filter_xx.go code, OSS and g3 will have different list.

with this list,  each filter decides if it needs to be added.  that is why moving the "if" checking to the filter config gen code is important. The high level code just blindly loop through each registered filter, to call its config_gen function.  IF return nil,  not to add it.

Teju has MakeFilterGenerator function, two separate one, one for OSS, and one for g3.

Reply:
I see, each filter_xx.go can statically register itself. Makes sense to me.

Reply:
My understanding about Wayne's proposal: 
- each filter register itself
- its config generator handles the logic if the filter is needed or not like checking `Skip_${FilterName}_Filter` flags
- OSS and g3 has two separate lists of targeted filter lists

My 2 cents: it would be overkill for our current requirement, swapping SC_V2 filter.

It is  will be confusing have two places to control the filter chain, one in the filter lists provided and another inside the filter config generator. I would prefer having a central sequential `MakeFilterGenerator` to control this,  and we don't necessarily need two `MakeFilterGenerator` but just patch the result it generates to replace SC_v2.

If we plan to have more filters in g3, we can easily refactor based on that.

Reply:
My idea is we don't have a high level function MakeFilterGenerator.   We just have a global variable registered_filter list.
By register or not register, we can control the list.  For example:
* we dont move in v1_service_control code in g3, so it will not be registered
* we don't move in backed_auth, path_rewrite into g3,  so they will not be in g3 list.
* v2_sc only in g3, not in OSS.  so their registered list will be different.

Technically, I am not sure if golang can do this easily.  If each filter is a package,  each package can have __init__funciton, that can register.  If each filter is a go file,  not sure if it can have its own __init__ function for each go file.

If it is too much trouble,  Having separate MakeFilterGenerator function is fine for me too

Reply:
Oh, I see. Good point of decoupling the config generation code from the whole library in terms of it is introduced in g3 or not! Though the readability of wrapping all the logic of if the filter config is generated into each package doesn't looks good, compared to a literal sequential `MakeFilterGenerator`, I am OK with that. 

Feasibility: yes, I know Golang should be able to do this.

Reply:
Not clear which approach you are going to take.   Are you going to add filter self-registration code?
If no,  I will review this code more carefully, and get this one in.

Reply:
Sorry for the misleading. I prefer not to add the registry for now.

Reply:
> I would prefer having a central sequential MakeFilterGenerator to control this, and we don't necessarily need two MakeFilterGenerator but just patch the result it generates to replace SC_v2.

I don't understand how this will work in g3. We need to support:
- One version of `MakeFilterGenerator` that will generate SC v1, backend auth, path rewrite, etc. for API Gateway.
- Another version of `MakeFilterGenerator` that will generate SC v2 for us.

I thought that the caller of config generator would decide the logic. But you're moving the decision to config generator itself. There is only 1 copy of config generator inside g3. So how can we have 2 different filter generation decisions?

Reply:
In the import target [doc](https://docs.google.com/document/d/1TDyjFNfkx-s2NaopWEK_-xYi1l1DVEIYYSZBlKYgPng/edit#bookmark=id.dnilp44edtew), I thought we got aligned about getting the filter config chains from ESPv2 lib and replacing the SC v2 filter, because we are essentially applying the GKE sidecar deployment model for CAP. 

Please correct me if I miss sth?

Reply:
Discussed offline, I understand the flow now. Thanks!